Improve kernels launch checks

pzehner opened the pull request:
### Context

This PR is loosely related to #8103, where it appeared that some kernels launch checks may only take place in debug mode, or specific debug mode (bounds check), may be redundant, or may be inaccurate (checks against wrong values).

I investigated and found that the following files are implied when checking the validity of a kernel before/during its launch (in the order of execution):

| Location | Check | Ref. value | Action | Build |
|-|-|-|-|-|
| [`KokkosExp_MDRangePolicy.hpp:421`](https://github.com/kokkos/kokkos/blob/ec1e636b19c646cff56b5f42dd59d0f2838741f9/core/src/KokkosExp_MDRangePolicy.hpp#L421) | Product of threads per block (aka product of tiles) | Wrong API value\* | Print, abort | Any |
| [`Kokkos_Cuda_Parallel_MDRange.hpp:118`](https://github.com/kokkos/kokkos/blob/ec1e636b19c646cff56b5f42dd59d0f2838741f9/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp#L118) | Threads per block (aka tiles) | API values | Abort | Debug only |
| [`Kokkos_Cuda_Parallel_MDRange.hpp:124`](https://github.com/kokkos/kokkos/blob/ec1e636b19c646cff56b5f42dd59d0f2838741f9/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp#L124) | Product of threads per block (aka product of tiles) | API values | Abort | Debug only |
| [`Kokkos_Cuda_Parallel_MDRange.hpp:131`](https://github.com/kokkos/kokkos/blob/ec1e636b19c646cff56b5f42dd59d0f2838741f9/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp#L131) | Blocks per grid (aka number of tiles) | API values | Abort | Debug only |
| [`Kokkos_Cuda_KernelLaunch.hpp:699`](https://github.com/kokkos/kokkos/blob/ec1e636b19c646cff56b5f42dd59d0f2838741f9/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp#L699) [`Kokkos_HIP_KernelLaunch.hpp:583`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp#L583) | Kernel was successfully launched (aka synchronous check) || Exception | Debug with bounds check only |
| [`Kokkos_Cuda_KernelLaunch.hpp:699`](https://github.com/kokkos/kokkos/blob/ec1e636b19c646cff56b5f42dd59d0f2838741f9/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp#L699) [`Kokkos_HIP_KernelLaunch.hpp:584`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp#L584) | Kernel was successfully executed (aka asynchronous check) || Fence, abort | Debug with bounds check only |

Almost all checks only take place on debug mode. Note that most checks are performed with the Cuda backend. The total number of threads per block may be checked twice (against different values).

\* The value depends on the backend:

| Backend | Current | Should be | Location |
|-|-|-|-|
| Cuda | `maxThreadsPerMultiProcessor` | `maxThreadsPerBlock` | [`Kokkos_Cuda_MDRangePolicy.hpp:43`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/Cuda/Kokkos_Cuda_MDRangePolicy.hpp#L43) |
| HIP | `maxThreadsPerSM` | `MaxThreadsPerBlock` | [`Kokkos_HIP_MDRangePolicy.hpp:43`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/HIP/Kokkos_HIP_MDRangePolicy.hpp#L43) |
| SYCL | `maxWorkgroupSize` | `maxWorkgroupSize` | [`Kokkos_SYCL_MDRangePolicy.hpp:44`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/SYCL/Kokkos_SYCL_MDRangePolicy.hpp#L44) |

Note that for Cuda and HIP, the current value is larger than the appropriate value, and the check may succeed when it should not.

In addition to that, `MDRangePolicy.max_total_tile_size()` gives the appropriate value, for all backends but for Cuda:

| Backend | Current | Location |
|-|-|-|
| Cuda | 512 | [`Kokkos_Cuda_MDRangePolicy.hpp:47`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/Cuda/Kokkos_Cuda_MDRangePolicy.hpp#L47) |
| HIP | `MaxThreadsPerBlock` | [`Kokkos_HIP_MDRangePolicy.hpp:46`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/HIP/Kokkos_HIP_MDRangePolicy.hpp#L46) |
| SYCL | `maxWorkgroupSize` | [`Kokkos_SYCL_MDRangePolicy.hpp:47`](https://github.com/kokkos/kokkos/blob/28848dfa8fc1411873b6a663c4d29ba05712510e/core/src/SYCL/Kokkos_SYCL_MDRangePolicy.hpp#L47) |

The values of the above two tables are actually stored in the `TileSizeProperties` struct, that differentiates `max_threads` and `max_total_tile_size` (along with other hardcoded values). I'm not sure this distinction is useful. Besides, `max_total_tile_size` is rarely used.

### Idea of the PR

I think we can rationalize these different checks:

- We can check number of threads per block/product of threads per block/blocks per grid in debug build, to have an explicit error message, for Cuda and HIP;
- Check if a kernel immediately failed to launch (i.e. synchronous errors) in any build;
- ~Do not check if a kernel failed to execute (i.e. asynchronous errors) anymore~; and
- Harmonize API values to be checked against.

~Validity of the kernel launch parameters are checked by the Cuda, HIP, and SYCL APIs respectively (but the outcome of this is only checked by Kokkos on debug mode with bounds check, for the former two). I'm not sure why does Kokkos have similar checks of its own? IS it to have better error messages?~

### Progress

- [x] Write tests that launch a kernel with obviously too large, slightly too large, or incorrect launch parameters;
- [x] Move tests to `core/unit_test/TestMDRangePolicyConstructors.hpp`;
- [ ] Check number of threads per block/product of threads per block/blocks per grid in debug build for Cuda and HIP and have meaningful messages (wait for #8264);
  - [ ] Cuda;
  - [ ] HIP;
  - [ ] SYCL;
- [ ] Check synchronous errors in any build and use `kokkos_abort`;
  - [x] Cuda;
  - [x] HIP;
  - [ ] SYCL;
- [ ] Harmonize API values to be checked against;
  - [ ] Cuda;
  - [ ] HIP;
  - [ ] SYCL.

pzehner committed:
Add tests to check limits for 4D

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

masterleinad commented:
> Validity of the kernel launch parameters are checked by the Cuda, HIP, and SYCL APIs respectively (but the outcome of this is only checked by Kokkos on debug mode with bounds check, for the former two). I'm not sure why does Kokkos have similar checks of its own? IS it to have better error messages?

`MDRangePolicy` or any other policy isn't specialized for a particular backend but uses properties from the respective backends to be set uop indirectly. Hence, it makes sense that setting up the tile sizes (providing sensible defaults) and checking for the validity happens in that class. Since this is somewhat general, it doesn't/can't take particularities for the mapping to the respective resource into account. Instead we perform those checks in the backend specializations for the respective parallel constructs. I agree that that means that the checks feel sonewhat duplicated but I think that design is OK. In any case, we should always do checks for tile sizes and launch configurations, not just in `Debug` mode.
On the other hand, checking for launch success reqiuires fencing so we shouldn't do that unless it's explicitly requested.

Separately, I agree that we have to fix `max_threads` for `Cuda` and `HIP`. `max_total_tile_size` might be a misnormer and should maybe be `max_default_tile_size` (assuming that's what we meant).

pzehner commented:
> On the other hand, checking for launch success reqiuires fencing so we shouldn't do that unless it's explicitly requested.

Are we sure about this? If Cuda cannot launch a kernel, shouldn't `cudaGetLastError` return the error code immediately after?

On the other hand, checking for execution success does require fencing, and for this I agree it shouldn't be implicitly enabled.

pzehner commented:
@masterleinad Following [this blog post](https://leimao.github.io/blog/Proper-CUDA-Error-Checking/), I think it's safe to always check `cudaGetLastError` immediately after a kernel launch, which would capture synchronous errors (without fencing). Asynchronous errors, that require fencing, could be checked only if a specific new debug flag is passed (e.g. `Kokkos_ENABLE_DEBUG_FORCE_SYNC`). What do you think about this approach?

masterleinad commented:
> @masterleinad Following [this blog post](https://leimao.github.io/blog/Proper-CUDA-Error-Checking/), I think it's safe to always check `cudaGetLastError` immediately after a kernel launch, which would capture synchronous errors (without fencing). Asynchronous errors, that require fencing, could be checked only if a specific new debug flag is passed (e.g. `Kokkos_ENABLE_DEBUG_FORCE_SYNC`). What do you think about this approach?

Yes, we could look into always calling `KOKKOS_IMPL_CUDA_SAFE_CALL(cudaGetLastError());` after a kernel launch and look into the overhead that adds.
On the other hand, we are already using `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` to check for asynchronous errors and people rely on that for debugging. I don't see an immediate need to introduce a different option even if the current name might not be quite spot-on.

pzehner commented:
For me, the option name `Kokkos_ENABLE_DEBUG_BOUNDS_CHECK` is very misleading when it comes to asynchronous errors. Moreover, this behavior is not documented (neither on the docs, nor in the description string in CMake).

We may add the new option, keep the behavior of the current one, and add a warning to encourage users to use the new one if they only want to check asynchronous errors.

pzehner commented:
After a discussion with @dalg24, it's OK to remove the asynchronous error check completely, as it should be handled by other `fence`s in the code. The user can still request a synchronous execution if the backend programming model allows it (e.g. the environnement variable `CUDA_LAUNCH_BLOCKING` for Cuda).

The PR could be a good opportunity to harmonize checks for the different backends. Everything may not be tackled in this PR.

Exceptions should be converted to `kokkos_abort`.

From this, I think that the following behavior could be enforced by this PR (minimal impact):

- Check number of threads per block/product of threads per block/blocks per grid in debug build, to have an explicit error message, for Cuda and HIP;
- Check synchronous errors in any build, use `kokkos_abort`;
- Do not check asynchronous errors anymore; and
- Harmonize API values to be checked against.

pzehner committed:
Move tests

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner committed:
Fix typo and test category

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner committed:
Fix how to get maximum number of threads per block

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner committed:
Enable limits test only on GPU

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner commented:
The threads per block check will be facilitated by #8264.

pzehner committed:
Always fail on synchronous error

Cuda calls Kokkos::abort;
HIP seemingly raises an exception;
SYCL probably raises an exception too.

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner commented:
Retest this please

pzehner commented:
Retest this please

pzehner committed:
Fix HIP code

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

masterleinad commented on core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:
```diff
@@ -695,11 +695,8 @@ struct CudaParallelLaunchImpl<
-#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
+      // check any launch error (synchronous check)
       KOKKOS_IMPL_CUDA_SAFE_CALL(cudaGetLastError());
-      cuda_instance->fence(
-          "Kokkos::Impl::launch_kernel: Debug Only Check for Execution Error");
```
I'm opposed to removing the fence here. That should be its own discussion.

pzehner replied:
This comes from a discussion I had with @dalg24, if I'm not wrong the idea behind this is (as I wrote above):

> it's OK to remove the asynchronous error check completely, as it should be handled by other `fence`s in the code. The user can still request a synchronous execution if the backend programming model allows it (e.g. the environnement variable `CUDA_LAUNCH_BLOCKING` for Cuda).

If I remember the discussion correctly, Damien was opposed to this fencing.

My initial idea was to keep it, but enable it with a new specific option.

We can discuss about it. Why do you want to keep it?

masterleinad replied:
> We can discuss about it. Why do you want to keep it?

Mostly beacuse this is known behavior and it's handy/useful. Of coourse, we are not very consistent across backends to fence when `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` is set either. I'm fine with introducing a separate CMake option (possibly just enabling with `DEBUG` depending on how expensive it turns out to be) but I would focus on improving the checks in this pull request and not removing existing features.

pzehner replied:
> Mostly beacuse this is known behavior and it's handy/useful.

As I said, this is not documented.

> (possibly just enabling with `DEBUG` depending on how expensive it turns out to be)

This option is already expensive, as it forces synchronous execution of kernels.

pzehner replied:
After discussion, we decided to keep the asynchronous check for now.

pzehner committed:
Fix SYCL test

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

Adrien-Tab commented:
> The threads per block check will be facilitated by #8264.

I split this PR into two, now it's in https://github.com/kokkos/kokkos/pull/8344

pzehner commented:
We had a discussion with @dalg24 and @masterleinad, and this is what came out:

- For the synchronous check: this is kept and the overhead should be checked;
- For the asynchronous check: this should be discussed by its own PR, so the current change of this PR regarding it will be reverted;
- Tests should be moved to a different place.

pzehner committed:
Partially revert "Always fail on synchronous error" and allow asynchronous check

This reverts commit e83aaf332b5884b4c02c3ce74d61c35e57e2b4ed.

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner committed:
Move tests to their dedicated file

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

dalg24 commented on core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:
```diff
@@ -658,8 +658,11 @@ struct CudaParallelLaunchImpl<
       base_t::invoke_kernel(driver, grid, block, shmem, cuda_instance);
 
-#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
+      // check any launch error (synchronous check)
```
@crtrott brought up that this was a conscious decision in the past because the `cudaGetLastError()` call costed a few microseconds.
I suppose the launch latency benchmarks should confirm that this is a bad idea.  (Please check and post the results in the description)
We could add comments in the code to capture this.

masterleinad replied:
We discussed that we want to keep everything as-is for now here (making sure we get consistent behavior across all backends and document it).

pzehner replied:
The first benchmarks I've run on a H100 do not give a definitive impact on performance.

Keeping this synchronous check in debug mode only will make more issues like #8103 to show up... I would suggest to introduce a flag to enable it (like `Kokkos_DEBUG_CHECK_KERNEL_LAUNCH`).

masterleinad replied:
That still means that users need to enable an extra flag, though.

pzehner replied:
Yes, but it would have a meaningful name and would be documented.

pzehner committed:
Improve invalid tiling test message check and enable it for Cuda and HIP only

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>

pzehner committed:
Fix maximum number of threads per block in TileSizeProperties for Cuda

Signed-off-by: Paul Zehner <paul.zehner@cea.fr>
